qlty-cloud/src/auth/auth_flow.rs

```diff
@@ -99,53 +76,91 @@ fn redirect_matches(expected: String, actual: String) -> bool {
     false
 }
 
-fn login_server(state: Arc<AppState>) -> Result<(Server, u16)> {
-    let listener = TcpListener::bind("127.0.0.1:0")?;
-    let port = listener.local_addr()?.port();
-    let server = HttpServer::new(move || {
-        App::new()
-            .app_data(Data::new(state.clone()))
-            .service(auth_flow)
-    })
-    .shutdown_timeout(0)
-    .workers(1)
-    .listen(listener)?
-    .run();
-    Ok((server, port))
-}
+fn run_handler(request: &Request, state: &AppState) -> Result<ResponseBox> {
+    let uri = request.url().parse::<Uri>()?;
+    let params: AuthFlowQueryParams;
+    if let Some(query) = uri.query() {
+        params = serde_querystring::from_str::<AuthFlowQueryParams>(query, ParseMode::UrlEncoded)
+            .with_context(|| "invalid parameter")?;
+    } else {
+        bail!("No query parameters found in request");
+    }
 
-pub fn launch_login_server(state: AppState) -> Result<u16> {
-    let handle = state.server_handle.clone();
-    let (server, port) = login_server(Arc::new(state))?;
+    if params.state != state.original_state {
+        bail!("State does not match original state");
+    }
 
-    handle.lock().unwrap().replace(server.handle());
+    if !redirect_matches(state.login_url.clone(), params.redirect_uri.clone()) {
+        bail!("Redirect URI does not match login URL");
+    }
 
-    thread::spawn(move || {
-        let _ = tokio::runtime::Builder::new_current_thread()
-            .enable_all()
-            .build()
-            .unwrap()
-            .block_on(server);
-    });
+    let token = Token::new(&state.credential_user);
+    if let Err(e) = token.set(&params.code) {
+        bail!("Failed to store auth token in credential storage: {}", e);
+    }
 
-    Ok(port)
+    let redirect = format!("Location: {}", params.redirect_uri)
+        .parse::<Header>()
+        .map_err(|_| anyhow!("Failed to generate redirect"))?;
+    Ok(Response::empty(307).with_header(redirect).boxed())
 }
 
-fn stop_login_server(state: Data<Arc<AppState>>) {
-    rt::spawn(async move {
-        tokio::time::sleep(Duration::from_millis(100)).await;
-        let handle = state.server_handle.lock().unwrap().take().unwrap();
-        handle.stop(false).await;
+pub fn launch_login_server(state: AppState) -> Result<ServerResponse> {
+    let server = Server::http("127.0.0.1:0");
+    if let Err(err) = server {
+        bail!("Failed to create server: {}", err);
```

Comment:
```
the method `with_context` exists for enum `Result<Server, Box<dyn Error + Send + Sync>>`, but its trait bounds were not satisfied
the following trait bounds were not satisfied:
`Box<dyn StdError + Send + std::marker::Sync>: anyhow::context::ext::StdError`
which is required by `Result<tiny_http::Server, Box<dyn StdError + Send + std::marker::Sync>>: anyhow::Context<tiny_http::Server, Box<dyn StdError + Send + std::marker::Sync>>`
```

This is typically the error you get when you try to manipulate this error object, since it's not a regular error and doesn't implement Sized. I think we can simplify by using `map_error` but that would still lose "context".